### DISCUSSION TITLE
Vale's generational references and region borrow checker

### DISCUSSION THREAD

**Discussion by jordiae** [modular/modular#461] [Ideas]
I was wondering whether the Mojo team/community has been following Vale's [1] approach to memory safety, based on generational references and a region borrow checker!

[1] https://vale.dev/

**Comment by lattner** [comment]
Yep, I've followed it.  It is currently experimental and adds non-trivial overhead to the runtime of a program, so it will be interesting to see how it bakes out and matures in Vale. For Mojo, we're sticking with somewhat more traditional implementation approaches.

This isn't to say "no, we will never do this": we need to get more experience with the planned lifetimes and other features, and if they don't work then we'd consider it.  That said, I can say that "it isn't on our radar and it looks like there is more research to be done before we'd seriously consider it".

**Reply by jordiae** [reply]
Makes sense! I was really curious about your opinion. Thanks.
